source/microbit/mprun.c

```diff
@@ -105,9 +118,11 @@ void mp_run(void) {
     readline_init0();
     microbit_init();
 
-    if (APPENDED_SCRIPT->header[0] == 'M' && APPENDED_SCRIPT->header[1] == 'P') {
+    if (hard_reboot && APPENDED_SCRIPT->header[0] == 'M' && APPENDED_SCRIPT->header[1] == 'P') {
```

Comment:
What we can do here is only execute the appended script, or main.py, if in "friendly REPL" mode.  If in "raw REPL" mode then don't execute the start up script.  This is how the pyboard works.  When you use the pyboard.py tool to upload/execute a script it will go into raw REPL mode so that the initial main.py is not executed.  I think that's how we should solve this problem.